aea/test_tools/test_protocol.py

```diff
@@ -0,0 +1,169 @@
+            self.perform_message_test(msg)
+
+    def test_messages_inconsistent(self) -> None:
+        """Run messages are inconsistent."""
+        for msg in self.build_inconsistent():
+            assert not msg._is_consistent()  # pylint: disable=protected-access # nosec
+
+    def test_messages_fail_to_encode_decode(self) -> None:
+        """Run messages are failing to encode and decode."""
+        msg = self.build_messages()[0]
+        with patch.object(msg.__class__.Performative, "__eq__", return_value=False):
+            with pytest.raises(
+                ValueError, match=f"Performative not valid: {msg.performative}"
+            ):
+                msg.serializer.encode(msg)
+
+        encoded_msg = msg.serializer.encode(msg)
+        with patch.object(msg.__class__.Performative, "__eq__", return_value=False):
+            with pytest.raises(
+                ValueError, match=f"Performative not valid: {msg.performative}"
+            ):
+                msg.serializer.decode(encoded_msg)
+
+    @abstractmethod
+    def build_messages(self) -> List[Message]:
+        """Build the messages to be used for testing."""
+
+    @abstractmethod
+    def build_inconsistent(self) -> List[Message]:
+        """Build inconsistent messages to be used for testing."""
+
+
+class BaseProtocolDialoguesTestCase(TestCase):
+    """Base class to test message construction for the protocol."""
+
+    __test__ = False
+    MESSAGE_CLASS: Type[Message] = Message
+    DIALOGUE_CLASS: Type[Dialogue] = Dialogue
+    DIALOGUES_CLASS: Type[Dialogues] = Dialogues
+    ROLE_FOR_THE_FIRST_MESSAGE: Dialogue.Role
+
+    def role_from_first_message(  # pylint: disable=unused-argument
+        self,
+        message: Message,
+        receiver_address: Address,
+    ) -> Dialogue.Role:
+        """Infer the role of the agent from an incoming/outgoing first message
+
+        :param message: an incoming/outgoing first message
+        :param receiver_address: the address of the receiving agent
+        :return: The role of the agent
+        """
+        return self.ROLE_FOR_THE_FIRST_MESSAGE
+
+    def make_dialogues_class(self) -> Type[Dialogues]:
+        """Make dialogues class with specific role."""
+
+        def new_init(self_: Dialogues, self_address: Address) -> None:
+            """New init function."""
+            self.DIALOGUES_CLASS.__init__(  # type: ignore # pylint: disable=no-value-for-parameter
+                self_,
+                self_address=self_address,
+                role_from_first_message=self.role_from_first_message,
+                dialogue_class=self.DIALOGUE_CLASS,
+            )
+
+        return type(
+            f"ForTest{self.DIALOGUES_CLASS.__name__}",
+            (self.DIALOGUES_CLASS,),
+            {"__init__": new_init},
+        )
+
+    @abstractmethod
+    def make_message_content(self) -> dict:
+        """Make a dict with message contruction content for dialogues.create."""
```

Comment:
precisely this you can infer from the `__annotations__` to construct a hypothesis strategy from.